populator-machinery/controller.go

```diff
@@ -562,98 +661,98 @@ func (c *controller) syncPvc(ctx context.Context, key, pvcNamespace, pvcName str
+					rawBlock = true
 				}
-			} else {
-				con.VolumeMounts = []corev1.VolumeMount{
-					{
-						Name:      populatorPodVolumeName,
-						MountPath: c.mountPath,
-					},
+
+				// Calculate the args for the populator pod
+				var args []string
+				args, err = c.populatorArgs(rawBlock, unstructured)
+				if err != nil {
+					return err
 				}
-			}
-			if waitForFirstConsumer {
-				pod.Spec.NodeName = nodeName
-			}
-			_, err = c.kubeClient.CoreV1().Pods(c.populatorNamespace).Create(ctx, pod, metav1.CreateOptions{})
-			if err != nil {
-				c.recorder.Eventf(pvc, corev1.EventTypeWarning, reasonPodCreationError, "Failed to create populator pod: %s", err)
-				return err
-			}
-			c.recorder.Eventf(pvc, corev1.EventTypeNormal, reasonPodCreationSuccess, "Populator started")
 
-			// If PVC' doesn't exist yet, create it
-			if pvcPrime == nil {
-				pvcPrime = &corev1.PersistentVolumeClaim{
+				// Make the pod
+				pod = &corev1.Pod{
 					ObjectMeta: metav1.ObjectMeta{
-						Name:      pvcPrimeName,
+						Name:      podName,
 						Namespace: c.populatorNamespace,
 					},
-					Spec: corev1.PersistentVolumeClaimSpec{
-						AccessModes:      pvc.Spec.AccessModes,
-						Resources:        pvc.Spec.Resources,
-						StorageClassName: pvc.Spec.StorageClassName,
-						VolumeMode:       pvc.Spec.VolumeMode,
-					},
+					Spec: makePopulatePodSpec(pvcPrimeName),
 				}
-				if waitForFirstConsumer {
-					pvcPrime.Annotations = map[string]string{
-						annSelectedNode: nodeName,
+				pod.Spec.Volumes[0].VolumeSource.PersistentVolumeClaim.ClaimName = pvcPrimeName
+				con := &pod.Spec.Containers[0]
+				con.Image = c.imageName
+				con.Args = args
+				if rawBlock {
+					con.VolumeDevices = []corev1.VolumeDevice{
+						{
+							Name:       populatorPodVolumeName,
+							DevicePath: c.devicePath,
+						},
+					}
+				} else {
+					con.VolumeMounts = []corev1.VolumeMount{
+						{
+							Name:      populatorPodVolumeName,
+							MountPath: c.mountPath,
+						},
 					}
 				}
-				_, err = c.kubeClient.CoreV1().PersistentVolumeClaims(c.populatorNamespace).Create(ctx, pvcPrime, metav1.CreateOptions{})
+				if waitForFirstConsumer {
+					pod.Spec.NodeName = nodeName
+				}
+				_, err = c.kubeClient.CoreV1().Pods(c.populatorNamespace).Create(ctx, pod, metav1.CreateOptions{})
 				if err != nil {
-					c.recorder.Eventf(pvc, corev1.EventTypeWarning, reasonPVCCreationError, "Failed to create populator PVC: %s", err)
+					c.recorder.Eventf(pvc, corev1.EventTypeWarning, reasonPodCreationError, "Failed to create populator pod: %s", err)
 					return err
 				}
-			}
+				c.recorder.Eventf(pvc, corev1.EventTypeNormal, reasonPodCreationSuccess, "Populator started")
 
-			// We'll get called again later when the pod exists
-			return nil
+				// We'll get called again later when the pod exists
+				return nil
+			}
+		} else {
+			operationStart, err := c.populate(ctx, params)
```

Comment:
What does operationStart  indicate? Why is err == nil not sufficient?

Reply:
Updated. I was originally thinking we should check if pvcPrime gets bind with pv in the populate function, but seems like move it out is better.